site/formatting.tsx

```diff
@@ -63,108 +66,222 @@ export const splitContentIntoSectionsAndColumns = (
+        }
+    }
+
+    class StandaloneFigureHandler extends AbstractHandler {
+        handle(el: CheerioElement, context: ColumnsContext) {
+            const $el = cheerioEl(el)
+            if (
+                FigureHandler.isFigure(el) &&
+                isElementFlushingColumns(el.nextSibling) &&
+                isColumnsEmpty(context.columns)
+            ) {
+                context.columns = getColumns(WP_ColumnStyle.StickyLeft)
+                context.columns.first.append($el)
+                flushAndResetColumns(context) // not strictly necessary since we know the next element flushes but keeps the abstraction clean
+                return null
+            }
+            return super.handle(el, context)
+        }
+    }
+
+    class FigureHandler extends AbstractHandler {
+        static isFigure(el: CheerioElement) {
+            const $el = cheerioEl(el)
+            return (
+                el.name === "figure" ||
+                el.name === "iframe" ||
+                // Temporary support for old chart iframes
+                el.name === "address" ||
+                $el.hasClass("wp-block-image") ||
+                $el.hasClass("tableContainer") ||
+                // Temporary support for non-Gutenberg iframes wrapped in wpautop's <p>
+                // Also catches older iframes (e.g. https://ourworldindata.org/food-per-person#world-map-of-minimum-and-average-dietary-energy-requirement-mder-and-ader)
+                $el.find("iframe").length !== 0 ||
+                // TODO: remove temporary support for pre-Gutenberg images and associated captions
+                el.name === "h6" ||
+                ($el.find("img").length !== 0 &&
+                    !$el.hasClass(PROMINENT_LINK_CLASSNAME) &&
+                    !$el.find(
+                        ".wp-block-owid-additional-information[data-variation='merge-left']"
+                    ))
+            )
+        }
+
+        handle(el: CheerioElement, context: ColumnsContext) {
+            const $el = cheerioEl(el)
+            if (FigureHandler.isFigure(el)) {
+                context.columns.last.append($el)
+                return null
+            }
+            return super.handle(el, context)
+        }
+    }
+
+    class DefaultHandler extends AbstractHandler {
+        handle(el: CheerioElement, context: ColumnsContext) {
+            const $el = cheerioEl(el)
+            // Move non-heading, non-image content to the left column
+            context.columns.first.append($el)
+            return null
+        }
+    }
+
+    const fullWidthHandler = new FullWidthHandler()
+
+    // Set up chain of responsibility pattern. Elements are being passed
+    // through the chain until a handler can process them.
+    fullWidthHandler
+        .setNext(new H4Handler())
+        .setNext(new SideBySideHandler())
+        .setNext(new StandaloneFigureHandler())
+        .setNext(new FigureHandler())
+        .setNext(new DefaultHandler())
```

Comment:
Just to make sure I understand correctly:

- For each element in a section, handlers are executed one by one in order. 
- It's the responsibility of the handler to figure out whether to **1)** apply some transformation, or **2)** do nothing, pass responsibility onto the next handler in the chain.
- A handler should never do both 1) and 2) – both apply a transformation and additionally let other handlers apply transformations. Or is there a case where we might want this? A concrete example: could both `StandaloneFigureHandler` and `FigureHandler` run on a single element?

Is all of that right?

Reply:
thanks for looking into this @danielgavrilov. Yes that's all correct. 

About the last point, handlers are indeed mutually exclusive: the first one to match returns and breaks the chain. It makes the order of handlers important, e.g. if we were to swap `StandaloneFigureHandler` and `FigureHandler`, `StandaloneFigureHandler` would never match.

The chain is built so that the first handler to match is the right and only one, this is the contract you sign when putting elements through the chain.

I believe a variation of the pattern might allow for overriding behaviour (e.g. express's middleware), but so far I don't see a case where we might need or want this. I can see how this would be different if the chain were to apply layers of subsequent transformations. In this case however, we want to find out which one of these handler is the most suited one, and by design it is the first one to match.

Reply:
Right, thanks! All makes sense. I think it might be helpful to have a short summary of these points in the code comment?

Reply:
yep, I reused your bullet points, I think they're very clear 👍